// yaml/browser/dist/schema/yaml-1.1/dist/bool.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.trueTag = exports.falseTag = void 0;

var _Scalar = require("../../nodes/Scalar.js");

function boolStringify(_ref, ctx) {
  var value = _ref.value,
      source = _ref.source;
  var boolObj = value ? trueTag : falseTag;
  if (source && boolObj.test.test(source)) return source;
  return value ? ctx.options.trueStr : ctx.options.falseStr;
}

var trueTag = {
  identify: function identify(value) {
    return value === true;
  },
  "default": true,
  tag: 'tag:yaml.org,2002:bool',
  test: /^(?:Y|y|[Yy]es|YES|[Tt]rue|TRUE|[Oo]n|ON)$/,
  resolve: function resolve() {
    return new _Scalar.Scalar(true);
  },
  stringify: boolStringify
};
exports.trueTag = trueTag;
var falseTag = {
  identify: function identify(value) {
    return value === false;
  },
  "default": true,
  tag: 'tag:yaml.org,2002:bool',
  test: /^(?:N|n|[Nn]o|NO|[Ff]alse|FALSE|[Oo]ff|OFF)$/i,
  resolve: function resolve() {
    return new _Scalar.Scalar(false);
  },
  stringify: boolStringify
};
exports.falseTag = falseTag;
