// @nodelib/fs.walk/out/readers/dist/async.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

Object.defineProperty(exports, "__esModule", {
  value: true
});

var events_1 = require("events");

var fsScandir = require("@nodelib/fs.scandir");

var fastq = require("fastq");

var common = require("./common");

var reader_1 = require("./reader");

var AsyncReader =
/*#__PURE__*/
function (_reader_1$default) {
  _inherits(AsyncReader, _reader_1$default);

  function AsyncReader(_root, _settings) {
    var _this;

    _classCallCheck(this, AsyncReader);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(AsyncReader).call(this, _root, _settings));
    _this._settings = _settings;
    _this._scandir = fsScandir.scandir;
    _this._emitter = new events_1.EventEmitter();
    _this._queue = fastq(_this._worker.bind(_assertThisInitialized(_this)), _this._settings.concurrency);
    _this._isFatalError = false;
    _this._isDestroyed = false;

    _this._queue.drain = function () {
      if (!_this._isFatalError) {
        _this._emitter.emit('end');
      }
    };

    return _this;
  }

  _createClass(AsyncReader, [{
    key: "read",
    value: function read() {
      var _this2 = this;

      this._isFatalError = false;
      this._isDestroyed = false;
      setImmediate(function () {
        _this2._pushToQueue(_this2._root, _this2._settings.basePath);
      });
      return this._emitter;
    }
  }, {
    key: "destroy",
    value: function destroy() {
      if (this._isDestroyed) {
        throw new Error('The reader is already destroyed');
      }

      this._isDestroyed = true;

      this._queue.killAndDrain();
    }
  }, {
    key: "onEntry",
    value: function onEntry(callback) {
      this._emitter.on('entry', callback);
    }
  }, {
    key: "onError",
    value: function onError(callback) {
      this._emitter.once('error', callback);
    }
  }, {
    key: "onEnd",
    value: function onEnd(callback) {
      this._emitter.once('end', callback);
    }
  }, {
    key: "_pushToQueue",
    value: function _pushToQueue(directory, base) {
      var _this3 = this;

      var queueItem = {
        directory: directory,
        base: base
      };

      this._queue.push(queueItem, function (error) {
        if (error !== null) {
          _this3._handleError(error);
        }
      });
    }
  }, {
    key: "_worker",
    value: function _worker(item, done) {
      var _this4 = this;

      this._scandir(item.directory, this._settings.fsScandirSettings, function (error, entries) {
        if (error !== null) {
          done(error, undefined);
          return;
        }

        var _iteratorNormalCompletion = true;
        var _didIteratorError = false;
        var _iteratorError = undefined;

        try {
          for (var _iterator = entries[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
            var entry = _step.value;

            _this4._handleEntry(entry, item.base);
          }
        } catch (err) {
          _didIteratorError = true;
          _iteratorError = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion && _iterator["return"] != null) {
              _iterator["return"]();
            }
          } finally {
            if (_didIteratorError) {
              throw _iteratorError;
            }
          }
        }

        done(null, undefined);
      });
    }
  }, {
    key: "_handleError",
    value: function _handleError(error) {
      if (this._isDestroyed || !common.isFatalError(this._settings, error)) {
        return;
      }

      this._isFatalError = true;
      this._isDestroyed = true;

      this._emitter.emit('error', error);
    }
  }, {
    key: "_handleEntry",
    value: function _handleEntry(entry, base) {
      if (this._isDestroyed || this._isFatalError) {
        return;
      }

      var fullpath = entry.path;

      if (base !== undefined) {
        entry.path = common.joinPathSegments(base, entry.name, this._settings.pathSegmentSeparator);
      }

      if (common.isAppliedFilter(this._settings.entryFilter, entry)) {
        this._emitEntry(entry);
      }

      if (entry.dirent.isDirectory() && common.isAppliedFilter(this._settings.deepFilter, entry)) {
        this._pushToQueue(fullpath, base === undefined ? undefined : entry.path);
      }
    }
  }, {
    key: "_emitEntry",
    value: function _emitEntry(entry) {
      this._emitter.emit('entry', entry);
    }
  }, {
    key: "isDestroyed",
    get: function get() {
      return this._isDestroyed;
    }
  }]);

  return AsyncReader;
}(reader_1["default"]);

exports["default"] = AsyncReader;
