// sucrase/dist/transformers/dist/ReactHotLoaderTransformer.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    "default": obj
  };
}

var _tokenizer = require('../parser/tokenizer');

var _Transformer = require('./Transformer');

var _Transformer2 = _interopRequireDefault(_Transformer);

var ReactHotLoaderTransformer =
/*#__PURE__*/
function (_Transformer2$default) {
  _inherits(ReactHotLoaderTransformer, _Transformer2$default);

  _createClass(ReactHotLoaderTransformer, [{
    key: "__init",
    value: function __init() {
      this.extractedDefaultExportName = null;
    }
  }]);

  function ReactHotLoaderTransformer(tokens, filePath) {
    var _this;

    _classCallCheck(this, ReactHotLoaderTransformer);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(ReactHotLoaderTransformer).call(this));
    _this.tokens = tokens;
    _this.filePath = filePath;

    ReactHotLoaderTransformer.prototype.__init.call(_assertThisInitialized(_this));

    ;
    return _this;
  }

  _createClass(ReactHotLoaderTransformer, [{
    key: "setExtractedDefaultExportName",
    value: function setExtractedDefaultExportName(extractedDefaultExportName) {
      this.extractedDefaultExportName = extractedDefaultExportName;
    }
  }, {
    key: "getPrefixCode",
    value: function getPrefixCode() {
      return "\n      (function () {\n        var enterModule = require('react-hot-loader').enterModule;\n        enterModule && enterModule(module);\n      })();".replace(/\s+/g, " ").trim();
    }
  }, {
    key: "getSuffixCode",
    value: function getSuffixCode() {
      var _this2 = this;

      var topLevelNames = new Set();
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = this.tokens.tokens[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var token = _step.value;

          if (!token.isType && _tokenizer.isTopLevelDeclaration.call(void 0, token) && token.identifierRole !== _tokenizer.IdentifierRole.ImportDeclaration) {
            topLevelNames.add(this.tokens.identifierNameForToken(token));
          }
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator["return"] != null) {
            _iterator["return"]();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      var namesToRegister = Array.from(topLevelNames).map(function (name) {
        return {
          variableName: name,
          uniqueLocalName: name
        };
      });

      if (this.extractedDefaultExportName) {
        namesToRegister.push({
          variableName: this.extractedDefaultExportName,
          uniqueLocalName: "default"
        });
      }

      return "\n;(function () {\n  var reactHotLoader = require('react-hot-loader').default;\n  var leaveModule = require('react-hot-loader').leaveModule;\n  if (!reactHotLoader) {\n    return;\n  }\n".concat(namesToRegister.map(function (_ref) {
        var variableName = _ref.variableName,
            uniqueLocalName = _ref.uniqueLocalName;
        return "  reactHotLoader.register(".concat(variableName, ", \"").concat(uniqueLocalName, "\", ").concat(JSON.stringify(_this2.filePath || ""), ");");
      }).join("\n"), "\n  leaveModule(module);\n})();");
    }
  }, {
    key: "process",
    value: function process() {
      return false;
    }
  }]);

  return ReactHotLoaderTransformer;
}(_Transformer2["default"]);

exports["default"] = ReactHotLoaderTransformer;
