// tailwindcss/src/cli/build/dist/deps.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loadPostcss = loadPostcss;
exports.loadPostcssImport = loadPostcssImport;
exports.loadCssNano = loadCssNano;
exports.loadAutoprefixer = loadAutoprefixer;

var _index = require("../../../peers/index.js");

// @ts-check
function loadPostcss() {
  // Try to load a local `postcss` version first
  try {
    return require('postcss');
  } catch (_unused) {}

  return (0, _index.lazyPostcss)();
}

function loadPostcssImport() {
  // Try to load a local `postcss-import` version first
  try {
    return require('postcss-import');
  } catch (_unused2) {}

  return (0, _index.lazyPostcssImport)();
}

function loadCssNano() {
  var options = {
    preset: ['default', {
      cssDeclarationSorter: false
    }]
  }; // Try to load a local `cssnano` version first

  try {
    return require('cssnano');
  } catch (_unused3) {}

  return (0, _index.lazyCssnano)()(options);
}

function loadAutoprefixer() {
  // Try to load a local `autoprefixer` version first
  try {
    return require('autoprefixer');
  } catch (_unused4) {}

  return (0, _index.lazyAutoprefixer)();
}
