// yaml/browser/dist/dist/index.dev.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Composer", {
  enumerable: true,
  get: function get() {
    return _composer.Composer;
  }
});
Object.defineProperty(exports, "Document", {
  enumerable: true,
  get: function get() {
    return _Document.Document;
  }
});
Object.defineProperty(exports, "Schema", {
  enumerable: true,
  get: function get() {
    return _Schema.Schema;
  }
});
Object.defineProperty(exports, "YAMLError", {
  enumerable: true,
  get: function get() {
    return _errors.YAMLError;
  }
});
Object.defineProperty(exports, "YAMLParseError", {
  enumerable: true,
  get: function get() {
    return _errors.YAMLParseError;
  }
});
Object.defineProperty(exports, "YAMLWarning", {
  enumerable: true,
  get: function get() {
    return _errors.YAMLWarning;
  }
});
Object.defineProperty(exports, "Alias", {
  enumerable: true,
  get: function get() {
    return _Alias.Alias;
  }
});
Object.defineProperty(exports, "isAlias", {
  enumerable: true,
  get: function get() {
    return _identity.isAlias;
  }
});
Object.defineProperty(exports, "isCollection", {
  enumerable: true,
  get: function get() {
    return _identity.isCollection;
  }
});
Object.defineProperty(exports, "isDocument", {
  enumerable: true,
  get: function get() {
    return _identity.isDocument;
  }
});
Object.defineProperty(exports, "isMap", {
  enumerable: true,
  get: function get() {
    return _identity.isMap;
  }
});
Object.defineProperty(exports, "isNode", {
  enumerable: true,
  get: function get() {
    return _identity.isNode;
  }
});
Object.defineProperty(exports, "isPair", {
  enumerable: true,
  get: function get() {
    return _identity.isPair;
  }
});
Object.defineProperty(exports, "isScalar", {
  enumerable: true,
  get: function get() {
    return _identity.isScalar;
  }
});
Object.defineProperty(exports, "isSeq", {
  enumerable: true,
  get: function get() {
    return _identity.isSeq;
  }
});
Object.defineProperty(exports, "Pair", {
  enumerable: true,
  get: function get() {
    return _Pair.Pair;
  }
});
Object.defineProperty(exports, "Scalar", {
  enumerable: true,
  get: function get() {
    return _Scalar.Scalar;
  }
});
Object.defineProperty(exports, "YAMLMap", {
  enumerable: true,
  get: function get() {
    return _YAMLMap.YAMLMap;
  }
});
Object.defineProperty(exports, "YAMLSeq", {
  enumerable: true,
  get: function get() {
    return _YAMLSeq.YAMLSeq;
  }
});
Object.defineProperty(exports, "Lexer", {
  enumerable: true,
  get: function get() {
    return _lexer.Lexer;
  }
});
Object.defineProperty(exports, "LineCounter", {
  enumerable: true,
  get: function get() {
    return _lineCounter.LineCounter;
  }
});
Object.defineProperty(exports, "Parser", {
  enumerable: true,
  get: function get() {
    return _parser.Parser;
  }
});
Object.defineProperty(exports, "parse", {
  enumerable: true,
  get: function get() {
    return _publicApi.parse;
  }
});
Object.defineProperty(exports, "parseAllDocuments", {
  enumerable: true,
  get: function get() {
    return _publicApi.parseAllDocuments;
  }
});
Object.defineProperty(exports, "parseDocument", {
  enumerable: true,
  get: function get() {
    return _publicApi.parseDocument;
  }
});
Object.defineProperty(exports, "stringify", {
  enumerable: true,
  get: function get() {
    return _publicApi.stringify;
  }
});
Object.defineProperty(exports, "visit", {
  enumerable: true,
  get: function get() {
    return _visit.visit;
  }
});
Object.defineProperty(exports, "visitAsync", {
  enumerable: true,
  get: function get() {
    return _visit.visitAsync;
  }
});
exports.CST = void 0;

var _composer = require("./compose/composer.js");

var _Document = require("./doc/Document.js");

var _Schema = require("./schema/Schema.js");

var _errors = require("./errors.js");

var _Alias = require("./nodes/Alias.js");

var _identity = require("./nodes/identity.js");

var _Pair = require("./nodes/Pair.js");

var _Scalar = require("./nodes/Scalar.js");

var _YAMLMap = require("./nodes/YAMLMap.js");

var _YAMLSeq = require("./nodes/YAMLSeq.js");

var cst = _interopRequireWildcard(require("./parse/cst.js"));

exports.CST = cst;

var _lexer = require("./parse/lexer.js");

var _lineCounter = require("./parse/line-counter.js");

var _parser = require("./parse/parser.js");

var _publicApi = require("./public-api.js");

var _visit = require("./visit.js");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
