// sucrase/dist/dist/register.dev.js
"use strict";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _interopRequireWildcard(obj) {
  if (obj && obj.__esModule) {
    return obj;
  } else {
    var newObj = {};

    if (obj != null) {
      for (var key in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, key)) {
          newObj[key] = obj[key];
        }
      }
    }

    newObj["default"] = obj;
    return newObj;
  }
}

var _pirates = require('pirates');

var pirates = _interopRequireWildcard(_pirates);

var _index = require('./index');

function addHook(extension, sucraseOptions, hookOptions) {
  var mergedSucraseOptions = sucraseOptions;
  var sucraseOptionsEnvJSON = process.env.SUCRASE_OPTIONS;

  if (sucraseOptionsEnvJSON) {
    mergedSucraseOptions = _objectSpread({}, mergedSucraseOptions, {}, JSON.parse(sucraseOptionsEnvJSON));
  }

  return pirates.addHook(function (code, filePath) {
    var _index$transform$call = _index.transform.call(void 0, code, _objectSpread({}, mergedSucraseOptions, {
      sourceMapOptions: {
        compiledFilename: filePath
      },
      filePath: filePath
    })),
        transformedCode = _index$transform$call.code,
        sourceMap = _index$transform$call.sourceMap;

    var mapBase64 = Buffer.from(JSON.stringify(sourceMap)).toString("base64");
    var suffix = "//# sourceMappingURL=data:application/json;charset=utf-8;base64,".concat(mapBase64);
    return "".concat(transformedCode, "\n").concat(suffix);
  }, _objectSpread({}, hookOptions, {
    exts: [extension]
  }));
}

exports.addHook = addHook;

function registerJS(hookOptions) {
  return addHook(".js", {
    transforms: ["imports", "flow", "jsx"]
  }, hookOptions);
}

exports.registerJS = registerJS;

function registerJSX(hookOptions) {
  return addHook(".jsx", {
    transforms: ["imports", "flow", "jsx"]
  }, hookOptions);
}

exports.registerJSX = registerJSX;

function registerTS(hookOptions) {
  return addHook(".ts", {
    transforms: ["imports", "typescript"]
  }, hookOptions);
}

exports.registerTS = registerTS;

function registerTSX(hookOptions) {
  return addHook(".tsx", {
    transforms: ["imports", "typescript", "jsx"]
  }, hookOptions);
}

exports.registerTSX = registerTSX;

function registerTSLegacyModuleInterop(hookOptions) {
  return addHook(".ts", {
    transforms: ["imports", "typescript"],
    enableLegacyTypeScriptModuleInterop: true
  }, hookOptions);
}

exports.registerTSLegacyModuleInterop = registerTSLegacyModuleInterop;

function registerTSXLegacyModuleInterop(hookOptions) {
  return addHook(".tsx", {
    transforms: ["imports", "typescript", "jsx"],
    enableLegacyTypeScriptModuleInterop: true
  }, hookOptions);
}

exports.registerTSXLegacyModuleInterop = registerTSXLegacyModuleInterop;

function registerAll(hookOptions) {
  var reverts = [registerJS(hookOptions), registerJSX(hookOptions), registerTS(hookOptions), registerTSX(hookOptions)];
  return function () {
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = reverts[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var fn = _step.value;
        fn();
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator["return"] != null) {
          _iterator["return"]();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
  };
}

exports.registerAll = registerAll;
