// yaml/dist/schema/dist/tags.dev.js
'use strict';

var map = require('./common/map.js');

var _null = require('./common/null.js');

var seq = require('./common/seq.js');

var string = require('./common/string.js');

var bool = require('./core/bool.js');

var _float = require('./core/float.js');

var _int = require('./core/int.js');

var schema = require('./core/schema.js');

var schema$1 = require('./json/schema.js');

var binary = require('./yaml-1.1/binary.js');

var omap = require('./yaml-1.1/omap.js');

var pairs = require('./yaml-1.1/pairs.js');

var schema$2 = require('./yaml-1.1/schema.js');

var set = require('./yaml-1.1/set.js');

var timestamp = require('./yaml-1.1/timestamp.js');

var schemas = new Map([['core', schema.schema], ['failsafe', [map.map, seq.seq, string.string]], ['json', schema$1.schema], ['yaml11', schema$2.schema], ['yaml-1.1', schema$2.schema]]);
var tagsByName = {
  binary: binary.binary,
  bool: bool.boolTag,
  "float": _float["float"],
  floatExp: _float.floatExp,
  floatNaN: _float.floatNaN,
  floatTime: timestamp.floatTime,
  "int": _int["int"],
  intHex: _int.intHex,
  intOct: _int.intOct,
  intTime: timestamp.intTime,
  map: map.map,
  "null": _null.nullTag,
  omap: omap.omap,
  pairs: pairs.pairs,
  seq: seq.seq,
  set: set.set,
  timestamp: timestamp.timestamp
};
var coreKnownTags = {
  'tag:yaml.org,2002:binary': binary.binary,
  'tag:yaml.org,2002:omap': omap.omap,
  'tag:yaml.org,2002:pairs': pairs.pairs,
  'tag:yaml.org,2002:set': set.set,
  'tag:yaml.org,2002:timestamp': timestamp.timestamp
};

function getTags(customTags, schemaName) {
  var tags = schemas.get(schemaName);

  if (!tags) {
    if (Array.isArray(customTags)) tags = [];else {
      var keys = Array.from(schemas.keys()).filter(function (key) {
        return key !== 'yaml11';
      }).map(function (key) {
        return JSON.stringify(key);
      }).join(', ');
      throw new Error("Unknown schema \"".concat(schemaName, "\"; use one of ").concat(keys, " or define customTags array"));
    }
  }

  if (Array.isArray(customTags)) {
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = customTags[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var tag = _step.value;
        tags = tags.concat(tag);
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator["return"] != null) {
          _iterator["return"]();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
  } else if (typeof customTags === 'function') {
    tags = customTags(tags.slice());
  }

  return tags.map(function (tag) {
    if (typeof tag !== 'string') return tag;
    var tagObj = tagsByName[tag];
    if (tagObj) return tagObj;
    var keys = Object.keys(tagsByName).map(function (key) {
      return JSON.stringify(key);
    }).join(', ');
    throw new Error("Unknown custom tag \"".concat(tag, "\"; use one of ").concat(keys));
  });
}

exports.coreKnownTags = coreKnownTags;
exports.getTags = getTags;
