// sucrase/dist/parser/util/dist/charcodes.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var charCodes;

(function (charCodes) {
  var backSpace = 8;
  charCodes[charCodes["backSpace"] = backSpace] = "backSpace";
  var lineFeed = 10;
  charCodes[charCodes["lineFeed"] = lineFeed] = "lineFeed"; //  '\n'

  var tab = 9;
  charCodes[charCodes["tab"] = tab] = "tab"; //  '\t'

  var carriageReturn = 13;
  charCodes[charCodes["carriageReturn"] = carriageReturn] = "carriageReturn"; //  '\r'

  var shiftOut = 14;
  charCodes[charCodes["shiftOut"] = shiftOut] = "shiftOut";
  var space = 32;
  charCodes[charCodes["space"] = space] = "space";
  var exclamationMark = 33;
  charCodes[charCodes["exclamationMark"] = exclamationMark] = "exclamationMark"; //  '!'

  var quotationMark = 34;
  charCodes[charCodes["quotationMark"] = quotationMark] = "quotationMark"; //  '"'

  var numberSign = 35;
  charCodes[charCodes["numberSign"] = numberSign] = "numberSign"; //  '#'

  var dollarSign = 36;
  charCodes[charCodes["dollarSign"] = dollarSign] = "dollarSign"; //  '$'

  var percentSign = 37;
  charCodes[charCodes["percentSign"] = percentSign] = "percentSign"; //  '%'

  var ampersand = 38;
  charCodes[charCodes["ampersand"] = ampersand] = "ampersand"; //  '&'

  var apostrophe = 39;
  charCodes[charCodes["apostrophe"] = apostrophe] = "apostrophe"; //  '''

  var leftParenthesis = 40;
  charCodes[charCodes["leftParenthesis"] = leftParenthesis] = "leftParenthesis"; //  '('

  var rightParenthesis = 41;
  charCodes[charCodes["rightParenthesis"] = rightParenthesis] = "rightParenthesis"; //  ')'

  var asterisk = 42;
  charCodes[charCodes["asterisk"] = asterisk] = "asterisk"; //  '*'

  var plusSign = 43;
  charCodes[charCodes["plusSign"] = plusSign] = "plusSign"; //  '+'

  var comma = 44;
  charCodes[charCodes["comma"] = comma] = "comma"; //  ','

  var dash = 45;
  charCodes[charCodes["dash"] = dash] = "dash"; //  '-'

  var dot = 46;
  charCodes[charCodes["dot"] = dot] = "dot"; //  '.'

  var slash = 47;
  charCodes[charCodes["slash"] = slash] = "slash"; //  '/'

  var digit0 = 48;
  charCodes[charCodes["digit0"] = digit0] = "digit0"; //  '0'

  var digit1 = 49;
  charCodes[charCodes["digit1"] = digit1] = "digit1"; //  '1'

  var digit2 = 50;
  charCodes[charCodes["digit2"] = digit2] = "digit2"; //  '2'

  var digit3 = 51;
  charCodes[charCodes["digit3"] = digit3] = "digit3"; //  '3'

  var digit4 = 52;
  charCodes[charCodes["digit4"] = digit4] = "digit4"; //  '4'

  var digit5 = 53;
  charCodes[charCodes["digit5"] = digit5] = "digit5"; //  '5'

  var digit6 = 54;
  charCodes[charCodes["digit6"] = digit6] = "digit6"; //  '6'

  var digit7 = 55;
  charCodes[charCodes["digit7"] = digit7] = "digit7"; //  '7'

  var digit8 = 56;
  charCodes[charCodes["digit8"] = digit8] = "digit8"; //  '8'

  var digit9 = 57;
  charCodes[charCodes["digit9"] = digit9] = "digit9"; //  '9'

  var colon = 58;
  charCodes[charCodes["colon"] = colon] = "colon"; //  ':'

  var semicolon = 59;
  charCodes[charCodes["semicolon"] = semicolon] = "semicolon"; //  ';'

  var lessThan = 60;
  charCodes[charCodes["lessThan"] = lessThan] = "lessThan"; //  '<'

  var equalsTo = 61;
  charCodes[charCodes["equalsTo"] = equalsTo] = "equalsTo"; //  '='

  var greaterThan = 62;
  charCodes[charCodes["greaterThan"] = greaterThan] = "greaterThan"; //  '>'

  var questionMark = 63;
  charCodes[charCodes["questionMark"] = questionMark] = "questionMark"; //  '?'

  var atSign = 64;
  charCodes[charCodes["atSign"] = atSign] = "atSign"; //  '@'

  var uppercaseA = 65;
  charCodes[charCodes["uppercaseA"] = uppercaseA] = "uppercaseA"; //  'A'

  var uppercaseB = 66;
  charCodes[charCodes["uppercaseB"] = uppercaseB] = "uppercaseB"; //  'B'

  var uppercaseC = 67;
  charCodes[charCodes["uppercaseC"] = uppercaseC] = "uppercaseC"; //  'C'

  var uppercaseD = 68;
  charCodes[charCodes["uppercaseD"] = uppercaseD] = "uppercaseD"; //  'D'

  var uppercaseE = 69;
  charCodes[charCodes["uppercaseE"] = uppercaseE] = "uppercaseE"; //  'E'

  var uppercaseF = 70;
  charCodes[charCodes["uppercaseF"] = uppercaseF] = "uppercaseF"; //  'F'

  var uppercaseG = 71;
  charCodes[charCodes["uppercaseG"] = uppercaseG] = "uppercaseG"; //  'G'

  var uppercaseH = 72;
  charCodes[charCodes["uppercaseH"] = uppercaseH] = "uppercaseH"; //  'H'

  var uppercaseI = 73;
  charCodes[charCodes["uppercaseI"] = uppercaseI] = "uppercaseI"; //  'I'

  var uppercaseJ = 74;
  charCodes[charCodes["uppercaseJ"] = uppercaseJ] = "uppercaseJ"; //  'J'

  var uppercaseK = 75;
  charCodes[charCodes["uppercaseK"] = uppercaseK] = "uppercaseK"; //  'K'

  var uppercaseL = 76;
  charCodes[charCodes["uppercaseL"] = uppercaseL] = "uppercaseL"; //  'L'

  var uppercaseM = 77;
  charCodes[charCodes["uppercaseM"] = uppercaseM] = "uppercaseM"; //  'M'

  var uppercaseN = 78;
  charCodes[charCodes["uppercaseN"] = uppercaseN] = "uppercaseN"; //  'N'

  var uppercaseO = 79;
  charCodes[charCodes["uppercaseO"] = uppercaseO] = "uppercaseO"; //  'O'

  var uppercaseP = 80;
  charCodes[charCodes["uppercaseP"] = uppercaseP] = "uppercaseP"; //  'P'

  var uppercaseQ = 81;
  charCodes[charCodes["uppercaseQ"] = uppercaseQ] = "uppercaseQ"; //  'Q'

  var uppercaseR = 82;
  charCodes[charCodes["uppercaseR"] = uppercaseR] = "uppercaseR"; //  'R'

  var uppercaseS = 83;
  charCodes[charCodes["uppercaseS"] = uppercaseS] = "uppercaseS"; //  'S'

  var uppercaseT = 84;
  charCodes[charCodes["uppercaseT"] = uppercaseT] = "uppercaseT"; //  'T'

  var uppercaseU = 85;
  charCodes[charCodes["uppercaseU"] = uppercaseU] = "uppercaseU"; //  'U'

  var uppercaseV = 86;
  charCodes[charCodes["uppercaseV"] = uppercaseV] = "uppercaseV"; //  'V'

  var uppercaseW = 87;
  charCodes[charCodes["uppercaseW"] = uppercaseW] = "uppercaseW"; //  'W'

  var uppercaseX = 88;
  charCodes[charCodes["uppercaseX"] = uppercaseX] = "uppercaseX"; //  'X'

  var uppercaseY = 89;
  charCodes[charCodes["uppercaseY"] = uppercaseY] = "uppercaseY"; //  'Y'

  var uppercaseZ = 90;
  charCodes[charCodes["uppercaseZ"] = uppercaseZ] = "uppercaseZ"; //  'Z'

  var leftSquareBracket = 91;
  charCodes[charCodes["leftSquareBracket"] = leftSquareBracket] = "leftSquareBracket"; //  '['

  var backslash = 92;
  charCodes[charCodes["backslash"] = backslash] = "backslash"; //  '\    '

  var rightSquareBracket = 93;
  charCodes[charCodes["rightSquareBracket"] = rightSquareBracket] = "rightSquareBracket"; //  ']'

  var caret = 94;
  charCodes[charCodes["caret"] = caret] = "caret"; //  '^'

  var underscore = 95;
  charCodes[charCodes["underscore"] = underscore] = "underscore"; //  '_'

  var graveAccent = 96;
  charCodes[charCodes["graveAccent"] = graveAccent] = "graveAccent"; //  '`'

  var lowercaseA = 97;
  charCodes[charCodes["lowercaseA"] = lowercaseA] = "lowercaseA"; //  'a'

  var lowercaseB = 98;
  charCodes[charCodes["lowercaseB"] = lowercaseB] = "lowercaseB"; //  'b'

  var lowercaseC = 99;
  charCodes[charCodes["lowercaseC"] = lowercaseC] = "lowercaseC"; //  'c'

  var lowercaseD = 100;
  charCodes[charCodes["lowercaseD"] = lowercaseD] = "lowercaseD"; //  'd'

  var lowercaseE = 101;
  charCodes[charCodes["lowercaseE"] = lowercaseE] = "lowercaseE"; //  'e'

  var lowercaseF = 102;
  charCodes[charCodes["lowercaseF"] = lowercaseF] = "lowercaseF"; //  'f'

  var lowercaseG = 103;
  charCodes[charCodes["lowercaseG"] = lowercaseG] = "lowercaseG"; //  'g'

  var lowercaseH = 104;
  charCodes[charCodes["lowercaseH"] = lowercaseH] = "lowercaseH"; //  'h'

  var lowercaseI = 105;
  charCodes[charCodes["lowercaseI"] = lowercaseI] = "lowercaseI"; //  'i'

  var lowercaseJ = 106;
  charCodes[charCodes["lowercaseJ"] = lowercaseJ] = "lowercaseJ"; //  'j'

  var lowercaseK = 107;
  charCodes[charCodes["lowercaseK"] = lowercaseK] = "lowercaseK"; //  'k'

  var lowercaseL = 108;
  charCodes[charCodes["lowercaseL"] = lowercaseL] = "lowercaseL"; //  'l'

  var lowercaseM = 109;
  charCodes[charCodes["lowercaseM"] = lowercaseM] = "lowercaseM"; //  'm'

  var lowercaseN = 110;
  charCodes[charCodes["lowercaseN"] = lowercaseN] = "lowercaseN"; //  'n'

  var lowercaseO = 111;
  charCodes[charCodes["lowercaseO"] = lowercaseO] = "lowercaseO"; //  'o'

  var lowercaseP = 112;
  charCodes[charCodes["lowercaseP"] = lowercaseP] = "lowercaseP"; //  'p'

  var lowercaseQ = 113;
  charCodes[charCodes["lowercaseQ"] = lowercaseQ] = "lowercaseQ"; //  'q'

  var lowercaseR = 114;
  charCodes[charCodes["lowercaseR"] = lowercaseR] = "lowercaseR"; //  'r'

  var lowercaseS = 115;
  charCodes[charCodes["lowercaseS"] = lowercaseS] = "lowercaseS"; //  's'

  var lowercaseT = 116;
  charCodes[charCodes["lowercaseT"] = lowercaseT] = "lowercaseT"; //  't'

  var lowercaseU = 117;
  charCodes[charCodes["lowercaseU"] = lowercaseU] = "lowercaseU"; //  'u'

  var lowercaseV = 118;
  charCodes[charCodes["lowercaseV"] = lowercaseV] = "lowercaseV"; //  'v'

  var lowercaseW = 119;
  charCodes[charCodes["lowercaseW"] = lowercaseW] = "lowercaseW"; //  'w'

  var lowercaseX = 120;
  charCodes[charCodes["lowercaseX"] = lowercaseX] = "lowercaseX"; //  'x'

  var lowercaseY = 121;
  charCodes[charCodes["lowercaseY"] = lowercaseY] = "lowercaseY"; //  'y'

  var lowercaseZ = 122;
  charCodes[charCodes["lowercaseZ"] = lowercaseZ] = "lowercaseZ"; //  'z'

  var leftCurlyBrace = 123;
  charCodes[charCodes["leftCurlyBrace"] = leftCurlyBrace] = "leftCurlyBrace"; //  '{'

  var verticalBar = 124;
  charCodes[charCodes["verticalBar"] = verticalBar] = "verticalBar"; //  '|'

  var rightCurlyBrace = 125;
  charCodes[charCodes["rightCurlyBrace"] = rightCurlyBrace] = "rightCurlyBrace"; //  '}'

  var tilde = 126;
  charCodes[charCodes["tilde"] = tilde] = "tilde"; //  '~'

  var nonBreakingSpace = 160;
  charCodes[charCodes["nonBreakingSpace"] = nonBreakingSpace] = "nonBreakingSpace"; // eslint-disable-next-line no-irregular-whitespace

  var oghamSpaceMark = 5760;
  charCodes[charCodes["oghamSpaceMark"] = oghamSpaceMark] = "oghamSpaceMark"; // ' '

  var lineSeparator = 8232;
  charCodes[charCodes["lineSeparator"] = lineSeparator] = "lineSeparator";
  var paragraphSeparator = 8233;
  charCodes[charCodes["paragraphSeparator"] = paragraphSeparator] = "paragraphSeparator";
})(charCodes || (exports.charCodes = charCodes = {}));

function isDigit(code) {
  return code >= charCodes.digit0 && code <= charCodes.digit9 || code >= charCodes.lowercaseA && code <= charCodes.lowercaseF || code >= charCodes.uppercaseA && code <= charCodes.uppercaseF;
}

exports.isDigit = isDigit;
